python/ray/tune/trial_runner.py

```diff
@@ -84,12 +92,94 @@ def __init__(self,
         self._global_time_limit = float(
             os.environ.get("TRIALRUNNER_WALLTIME_LIMIT", float('inf')))
         self._total_time = 0
+        self._iteration = 0
+        self._verbose = verbose
+        self._queue_trials = queue_trials
+
         self._server = None
+        self._server_port = server_port
         if launch_web_server:
-            self._server = TuneServer(self, server_port)
+            self._server = TuneServer(self, self._server_port)
+
+        self._trials = []
         self._stop_queue = []
-        self._verbose = verbose
-        self._queue_trials = queue_trials
+        self._checkpoint_dir = checkpoint_dir
+        self._checkpoint_mode = checkpoint_mode
+
+    def checkpoint(self):
+        """Saves execution state to `self._checkpoint_dir` if provided."""
+        if not self._checkpoint_mode or not self._checkpoint_dir:
+            return
+        checkpoint_dir = self._checkpoint_dir
+        if not os.path.exists(checkpoint_dir):
+            logger.debug("Checkpoint directory newly created.")
+            logger.warning("Search Algorithm and Scheduler not checkpointed.")
+            os.makedirs(checkpoint_dir)
+        runner_state = {
+            "checkpoints": list(
+                self.trial_executor.get_checkpoints().values()),
```

Comment:
By the way if you look at the pretty printed json, this ends up as a string somehow, and isn't very readable. The whole thing should be json.

Reply:
Fixed (this is because json was called twice)